File: leetcode/002-add-two-numbers/code.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if (l1 == nullptr) {
            return l2;
        }
        if (l2 == nullptr) {
            return l1;
        }

        auto p1 = l1;
        auto p2 = l2;
        for (;;) {
            if (p1 == nullptr && p2 == nullptr) {
                break;
            } else if (p1->next == nullptr && p2->next != nullptr) {
                p1->next = new ListNode;
            } else if (p1->next != nullptr && p2->next == nullptr) {
                p2->next = new ListNode;
            }
            p1 = p1->next;
            p2 = p2->next;
        }

        p1 = l1;
        p2 = l2;
        ListNode* tmp;
        int forward = 0;
        while (p1 != nullptr && p2 != nullptr) {
            auto sum = p1->val + p2->val + forward;
            if (sum >= 10) {
                p1->val = sum % 10;
                forward = sum / 10;
            } else {
                p1->val = sum;
                forward = 0;
            }
            if(p1->next == nullptr) {
                tmp = p1;
            }
            p1 = p1->next;
            p2 = p2->next;
        }
        if (forward != 0) {
            tmp->next = new ListNode;
            tmp->next->val = forward;
        }

        return l1;
    }
};

File: leetcode/002-add-two-numbers/code.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	} else if l2 == nil {
		return l1
	}

	// pad to the same length
	p1, p2 := l1, l2
	for {
		if p1.Next == nil && p2.Next == nil {
			break
		} else if p1.Next == nil && p2.Next != nil {
			p1.Next = new(ListNode)
			p1.Next.Val, p1.Next.Next = 0, nil
		} else if p1.Next != nil && p2.Next == nil {
			p2.Next = new(ListNode)
			p2.Next.Val, p2.Next.Next = 0, nil
		}
		p1, p2 = p1.Next, p2.Next
	}

	p1, p2 = l1, l2
	head := new(ListNode)
	head.Next = nil
	head.Val = 0
	forward := 0
	for p1 != nil && p2 != nil {
		sum := p1.Val + p2.Val + forward
		if sum >= 10 {
			forward, p1.Val = sum/10, sum%10
		} else {
			p1.Val, forward = sum, 0
		}
		p1, p2 = p1.Next, p2.Next
	}

	if forward != 0 {
		tmp := l1
		for tmp.Next != nil {
			tmp = tmp.Next
		}
		head.Val, tmp.Next = forward, head
	}

	return l1
}